### PR TITLE
Handle expired tokens

### PR DESCRIPTION
**Pull request by vargaat**
### What's new?
- Expired refresh tokens were not handled by the app. When the app was running and tokens expired all the API calls were failing. If the app was started with expired refresh token we just logged out the user.
- I refactored the already existing access token refresh mechanism and extended it with the re-login feature in case the refresh token is also expired.
- Refresh flow:
  - Expired access token is detected during an API call.
  - We are not calling the API request's handler with a failure.
  - Start token refresh process. New API calls made during this time are collected into a queue.
  - If access token is refreshed:
    - Execute queued requests.
    - Retry the original API call with the new token.
  - If access token refresh failed due to invalid refresh token:
    - Show a dialog informing the user about the need of re-login.
    - Show the login webview and catch the new token.
    - Execute queued requests.
    - Retry the original API call with the new token.

### How to test?
- There are two new options in the dev menu to invalidate the access and refresh tokens locally. We can't really use this to test the refresh token flow in the app because as soon as the dev menu is dismissed the tab bar tries a refresh that'll trigger the flow instantly. We can still test the app startup flow by killing the app while the dev menu is on screen and test the flow on the tab bar on which the dev menu is opened. You can invalidate only the access token to check that flow alone.
- Another option is to log in to the web with the user, go to your profile and delete the token your app created. This is a good way to randomly invalidate the token in the middle of any flow in the app.
- The 3rd option is to setup the automatic token invalidation plugin (see the epic for the link) and wait until the token is invalidated on server side. Unfortunately, this is 2 day at minimum.
- I think the main usage of this feature is when the token expires while the user is not actively using the app, then starts the app and it asks for re-login, like after a weekend.
- If the user cancels a flow or logs in with a different user then it's considered as a logout.

refs: [MBL-17417](https://instructure.atlassian.net/browse/MBL-17417)
affects: Student, Teacher, Parent
release note: Added re-login support for expired sessions.

## Screenshots

<table>
<tr>
<td></td>
<td></td>
</tr>
</table>

## Checklist

- [ ] Tested on phone
- [ ] Tested on tablet
- [ ] Approve from product

### CODE REVIEW

**Comment by inst-danger** [comment]
<details><summary>Teacher Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<details><summary>Parent Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

**Review by ndrsszsz — DISMISSED** [review]
Very nice solution. Plus points for the explaining comments here and there. 👍 
QA +1 (I didn't test all scenarios but it seems to be working good for me)

**Review by rh12 — APPROVED** [review]
QA + 1 (all apps, iPad @ iOS 17)

I found one edge case (may not be that important):
1. User is on the login screen with previous login listed
2. Token expires
3. User selects itself from list to relogin
4. Taps Cancel
App doesn't load the login screen, but shows the spinning canvas wheel forever

Some possible improvements:
- When tapping Cancel in during relogin the user is thrown back to the login screen, but without the current institution available as a button. If possible it would be nice to still show the button
  - this may be related to the unrelated, currently weird behaviour of the login buttons
  - this may not be possible currently if the logout upon cancel clears that button (not sure it does or doesn't)
- The dialog may be better with a "Please log in again" title and the current text or similar in a subtitle.

**Comment by vargaat** [comment]
I fixed the never ending loading spinner case, nice find! I'll keep your other improvement suggestions for another time now.

**Review by rh12 — APPROVED** [review]
Thanks for the quick fix!

File: Core/Core/Features/TokenRefresh/Model/LoginAgainInteractor.swift [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    func loginAgainOnExpiredRefreshToken(
+        tokenRefreshError: Error,
+        api: API
+    ) throws -> AnyPublisher<LoginSession, LoginError> {
+        guard
+            let host = api.loginSession?.baseURL.host(percentEncoded: false),
+            let rootViewController = AppEnvironment.shared.topViewController,
+            tokenRefreshError as? AccessTokenRefreshInteractor.TokenError == .expiredRefreshToken
```
[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/Features/TokenRefresh/Model/LoginAgainInteractor.swift:42]: It was a kind of hard for me to read.

**Review by suhaibabsi-inst — COMMENTED** [review]
Code +1

**Review by suhaibabsi-inst — APPROVED** [review]
QA +1

All apps, iPhone XS, iOS 18.3.0

File: Core/Core/Features/TokenRefresh/Model/LoginAgainInteractor.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (rh12) [reply Core/Core/Features/TokenRefresh/Model/LoginAgainInteractor.swift:42]: I agree. Either the parentheses, or we could add it to the error extensions like the other one

[the remark this answers is not in the document]
Comment (vargaat) [reply Core/Core/Features/TokenRefresh/Model/LoginAgainInteractor.swift:42]: I'll extract this into a helper in a follow-up PR.

File: Core/Core/Features/TokenRefresh/Model/TokenRefreshInteractor.swift [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        mainThread: AnySchedulerOf<DispatchQueue> = .main
+    ) {
+        self.api = api
+        self.accessTokenRefreshInteractor = accessTokenRefreshInteractor
+        self.waitingRequestsQueue.isSuspended = true
+        self.loginAgainInteractor = loginAgainInteractor
+        self.mainThread = mainThread
+    }
+
+    func isTokenRefreshInProgress() -> Bool {
```
[the remark this answers is not in the document]
Comment (vargaat) [reply Core/Core/Features/TokenRefresh/Model/TokenRefreshInteractor.swift:51]: I found it harder to manage the synchronous access (the getter and setter running on the `synchronizer` queue) without causing deadlocks.
